[Merged by Bors] - feat: several results about Monoid.Exponent

Ruben-VandeVelde opened the pull request:
---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

YaelDillies commented on Mathlib/GroupTheory/Exponent.lean:
```diff
@@ -287,11 +294,10 @@ theorem exponent_eq_zero_iff_range_orderOf_infinite (h : ∀ g : G, 0 < orderOf
-  refine' exponent_dvd_of_forall_pow_eq_one G _ fun g => _
-  obtain ⟨m, hm⟩ : orderOf g ∣ Finset.univ.lcm orderOf := Finset.dvd_lcm (Finset.mem_univ g)
-  rw [hm, pow_mul, pow_orderOf_eq_one, one_pow]
+theorem lcm_order_eq_exponent [Fintype G] : (Finset.univ : Finset G).lcm orderOf = exponent G :=
```
```suggestion
theorem lcm_orderOf_eq_exponent [Fintype G] : (Finset.univ : Finset G).lcm orderOf = exponent G :=
```
while you're at it

YaelDillies commented on Mathlib/GroupTheory/Exponent.lean:
```diff
@@ -106,8 +106,15 @@ theorem exponentExists_iff_ne_zero : ExponentExists G ↔ exponent G ≠ 0 := by
 @[to_additive]
-theorem exponent_eq_zero_iff : exponent G = 0 ↔ ¬ExponentExists G := by
-  simp only [exponentExists_iff_ne_zero, Classical.not_not]
+alias ⟨ExponentExists.exponent_ne_zero, _⟩ := exponentExists_iff_ne_zero
+
+@[to_additive]
+theorem ExponentExists.exponent_pos (h : ExponentExists G) : 0 < exponent G :=
+  h.exponent_ne_zero.bot_lt
+
+@[to_additive]
+theorem exponent_eq_zero_iff : exponent G = 0 ↔ ¬ExponentExists G :=
+  exponentExists_iff_ne_zero.symm.not_right
```
Can you
* turn around `exponentExists_iff_ne_zero`
* write `exponent_pos : 0 < exponent G ↔ ExponentExists G`
* derive `protected alias ⟨_, ExponentExists.exponent_pos⟩ := exponent_pos`

YaelDillies commented on Mathlib/GroupTheory/Exponent.lean:
```diff
@@ -178,25 +183,45 @@ theorem order_dvd_exponent (g : G) : orderOf g ∣ exponent G :=
 #align monoid.order_dvd_exponent Monoid.order_dvd_exponent
 #align add_monoid.add_order_dvd_exponent AddMonoid.addOrder_dvd_exponent
 
+@[to_additive addOrder_le_exponent]
+theorem order_le_exponent (h : ExponentExists G) (g : G) : orderOf g ≤ exponent G :=
```
```suggestion
theorem order_le_exponent (h : ExponentExists G) (g : G) : orderOf g ≤ exponent G :=
```
```suggestion
@[to_additive addOrderOf_le_exponent]
theorem orderOf_le_exponent (h : ExponentExists G) (g : G) : orderOf g ≤ exponent G :=
```

YaelDillies commented on Mathlib/GroupTheory/Exponent.lean:
```diff
@@ -178,25 +183,45 @@ theorem order_dvd_exponent (g : G) : orderOf g ∣ exponent G :=
 @[to_additive]
-theorem exponent_dvd_of_forall_pow_eq_one (G) [Monoid G] (n : ℕ) (hG : ∀ g : G, g ^ n = 1) :
-    exponent G ∣ n := by
+theorem exponent_dvd_iff_forall_pow_eq_one {n : ℕ} : exponent G ∣ n ↔ (∀ g : G, g ^ n = 1) := by
```
```suggestion
theorem exponent_dvd_iff_forall_pow_eq_one {n : ℕ} : exponent G ∣ n ↔ ∀ g : G, g ^ n = 1 := by
```

YaelDillies commented on Mathlib/GroupTheory/Exponent.lean:
```diff
@@ -178,25 +183,45 @@ theorem order_dvd_exponent (g : G) : orderOf g ∣ exponent G :=
+      exact hG
+    exact h₂.not_le h₃
+
+@[to_additive]
+theorem exponent_dvd_of_forall_pow_eq_one (n : ℕ) (hG : ∀ g : G, g ^ n = 1) :
+    exponent G ∣ n :=
+  (exponent_dvd_iff_forall_pow_eq_one G).mpr hG
```
Can you remove this one now that we have the iff?

YaelDillies commented on Mathlib/GroupTheory/Exponent.lean:
```diff
@@ -178,25 +183,45 @@ theorem order_dvd_exponent (g : G) : orderOf g ∣ exponent G :=
+theorem exponent_dvd {n : ℕ} : exponent G ∣ n ↔ ∀ g : G, orderOf g ∣ n := by
+  simp_rw [exponent_dvd_iff_forall_pow_eq_one, orderOf_dvd_iff_pow_eq_one]
+
+@[to_additive]
+theorem exponent_dvd_of_forall_orderOf_dvd (n : ℕ) (h : ∀ g : G, orderOf g ∣ n) : exponent G ∣ n :=
+  (exponent_dvd G).mpr h
```
Similarly here

YaelDillies commented on Mathlib/GroupTheory/Exponent.lean:
```diff
@@ -302,9 +328,17 @@ section LeftCancelMonoid
-theorem exponent_ne_zero_of_finite [Finite G] : exponent G ≠ 0 := by
-  cases nonempty_fintype G
-  simpa [← lcm_order_eq_exponent, Finset.lcm_eq_zero_iff] using fun x => (orderOf_pos x).ne'
+theorem ExponentExists.of_finite [Finite G] : Monoid.ExponentExists G := by
```
```suggestion
theorem ExponentExists.of_finite [Finite G] : ExponentExists G := by
```

YaelDillies commented on Mathlib/GroupTheory/Exponent.lean:
```diff
@@ -302,9 +328,17 @@ section LeftCancelMonoid
+
+@[to_additive]
+theorem exponent_ne_zero_of_finite [Finite G] : exponent G ≠ 0 :=
+  exponentExists_iff_ne_zero.mp ExponentExists.of_finite
```
```suggestion
  ExponentExists.of_finite.exponent_ne_zero
```

YaelDillies commented on Mathlib/GroupTheory/Exponent.lean:
```diff
@@ -315,18 +349,14 @@ section CommMonoid
-  rw [iSup]
-  rcases eq_or_ne (exponent G) 0 with (he | he)
-  · rw [he, Set.Infinite.Nat.sSup_eq_zero <| (exponent_eq_zero_iff_range_orderOf_infinite h).1 he]
+theorem exists_orderOf_eq_exponent (h : ∀ g : G, 0 < orderOf g) (he : exponent G ≠ 0) :
```
```suggestion
theorem exists_orderOf_eq_exponent (hG : ExponentExists G) :
```

YaelDillies commented on Mathlib/GroupTheory/Exponent.lean:
```diff
@@ -494,3 +546,77 @@ theorem Monoid.exponent_prod {M₁ M₂ : Type*} [Monoid M₁] [Monoid M₂] :
+  ⟨by rintro hx rfl; norm_num at hx, orderOf_eq_prime (hG ▸ Monoid.pow_exponent_eq_one x)⟩
+
+@[to_additive]
+theorem mul_comm_of_orderOf_dvd_two [IsCancelMul G] {a b : G} (h : ∀ g : G, orderOf g ∣ 2) :
+    a * b = b * a := by
```
```suggestion
theorem Commute.of_orderOf_dvd_two [IsCancelMul G] (h : ∀ g : G, orderOf g ∣ 2) :
    ∀ a b : G, Commute a b := by
```

YaelDillies commented on Mathlib/GroupTheory/Exponent.lean:
```diff
@@ -494,3 +546,77 @@ theorem Monoid.exponent_prod {M₁ M₂ : Type*} [Monoid M₁] [Monoid M₂] :
+    x⁻¹ = x :=
+  inv_eq_of_mul_eq_one_left <| pow_two (a := x) ▸ hx ▸ pow_orderOf_eq_one x
+
+/-- Any group of exponent two is abelian. -/
+@[to_additive (attr := reducible) "Any additive group of exponent two is abelian."]
+def commGroupOfExponentTwo (hG : Monoid.exponent G = 2) : CommGroup G where
+  mul_comm := mul_comm_of_exponent_two hG
```
I feel like you should not have this definition at all since it's just a call to a constructor with only one use of each argument

EDIT: I see this was moved. Can you put a `TODO: Delete` then?

YaelDillies commented on Mathlib/Algebra/Group/Commute/Defs.lean:
```diff
@@ -44,6 +44,9 @@ section Mul
 
 variable {S : Type*} [Mul S]
 
+@[to_additive]
+theorem _root_.commute_def {a b : S} : Commute a b ↔ a * b = b * a := .rfl
```
That's just the rewriting lemma for `Commute`, right?

Ruben-VandeVelde replied:
No, you also need to unfold `SemiconjBy`

adri326 replied:
You can use a `Commute a b` inside of a `rw`, but you can't easily `rw` in a `Commute a b` without first unfolding `Commute` *and* `SemiConjBy`

YaelDillies commented on Mathlib/GroupTheory/Exponent.lean:
```diff
@@ -314,7 +322,7 @@ theorem exponent_eq_zero_iff_range_orderOf_infinite (h : ∀ g : G, 0 < orderOf
   Nat.dvd_antisymm
     (lcm_orderOf_dvd_exponent G)
-    (exponent_dvd_of_forall_orderOf_dvd G _ fun g => Finset.dvd_lcm (Finset.mem_univ g))
+    ((exponent_dvd G).mpr fun g => Finset.dvd_lcm (Finset.mem_univ g))
```
`G` should be implicit in `exponent_dvd` since it's an iff

YaelDillies commented on Mathlib/GroupTheory/Exponent.lean:
```diff
@@ -349,8 +357,9 @@ section CommMonoid
 
 @[to_additive]
-theorem exists_orderOf_eq_exponent (h : ∀ g : G, 0 < orderOf g) (he : exponent G ≠ 0) :
+theorem exists_orderOf_eq_exponent (h : ∀ g : G, 0 < orderOf g) (hG : ExponentExists G) :
```
This first condition is redundant
```suggestion
theorem exists_orderOf_eq_exponent (hG : ExponentExists G) :
```

Ruben-VandeVelde replied:
Only if you assume cancellativity, as far as I can tell. If not, please prove :)

j-loreaux replied:
This is true in *any* monoid, almost by definition. I would put these right before `Monoid.exponent_ne_zero`.
```lean
@[to_additive]
theorem ExponentExists.isOfFinOrder (h : ExponentExists G) (g : G) : IsOfFinOrder g :=
  isOfFinOrder_iff_pow_eq_one.mpr <| by peel 2 h; exact this g

@[to_additive]
theorem ExponentExists.orderOf_pos (h : ExponentExists G) (g : G) : 0 < orderOf g :=
  h.isOfFinOrder g |>.orderOf_pos
```
you can use the latter to golf this proof too:
```lean
theorem exponent_eq_zero_of_order_zero {g : G} (hg : orderOf g = 0) : exponent G = 0 :=
  exponent_eq_zero_iff.mpr fun h ↦ h.orderOf_pos g |>.ne' hg
```

j-loreaux commented on Mathlib/GroupTheory/Exponent.lean:
```diff
@@ -178,25 +189,46 @@ theorem order_dvd_exponent (g : G) : orderOf g ∣ exponent G :=
+      exact hG
+    exact h₂.not_le h₃
+
+@[to_additive (attr := deprecated)]
+theorem exponent_dvd_of_forall_pow_eq_one (n : ℕ) (hG : ∀ g : G, g ^ n = 1) :
+    exponent G ∣ n :=
```
I think it would be nice to keep this around without the deprecation
```suggestion
@[to_additive]
theorem exponent_dvd_of_forall_pow_eq_one (n : ℕ) (hG : ∀ g : G, g ^ n = 1) :
    exponent G ∣ n :=
```

YaelDillies replied:
Can we make it an alias then?

j-loreaux reviewed (commented):
Yaël is correct, I have provided the proofs. Please remove the extraneous hypothesis. Otherwise LGTM.

adri326 commented on Mathlib/Algebra/Group/Commute/Defs.lean:
```diff
@@ -44,6 +44,9 @@ section Mul
 
 variable {S : Type*} [Mul S]
 
+@[to_additive]
+theorem _root_.commute_def {a b : S} : Commute a b ↔ a * b = b * a := .rfl
```
This lemma got merged in-between: [commute_iff_eq](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Group/Commute/Defs.html#commute_iff_eq)

YaelDillies commented on Mathlib/GroupTheory/Torsion.lean:
```diff
@@ -138,15 +138,15 @@ theorem ExponentExists.isTorsion (h : ExponentExists G) : IsTorsion G := fun g =
 @[to_additive is_add_torsion_of_finite "Finite additive groups are additive torsion groups."]
 theorem isTorsion_of_finite [Finite G] : IsTorsion G :=
-  ExponentExists.isTorsion <| exponentExists_iff_ne_zero.mpr exponent_ne_zero_of_finite
+  ExponentExists.isTorsion <| exponent_ne_zero.mp exponent_ne_zero_of_finite
```
```suggestion
  ExponentExists.of_finite.isTorsion
```

YaelDillies replied:
Why not use dot notation all the way?

Ruben-VandeVelde replied:
`_root_.ExponentExists.isTorsion` doesn't work in dot notation because it's about `Monoid.ExponentExists`. (I don't want to deal with that in this PR.)

YaelDillies replied:
Oh yeah that's why

YaelDillies commented on Mathlib/GroupTheory/Exponent.lean:
```diff
@@ -95,6 +95,14 @@ theorem _root_.MulOpposite.exponent : exponent (MulOpposite G) = exponent G := b
   all_goals exact ⟨(op_injective <| · <| op ·), (unop_injective <| . <| unop .)⟩
 
+@[to_additive]
+theorem ExponentExists.isOfFinOrder (h : ExponentExists G) (g : G) : IsOfFinOrder g :=
```
I'm willing to bet this is more useful in practice:
```suggestion
theorem ExponentExists.isOfFinOrder (h : ExponentExists G) {g : G} : IsOfFinOrder g :=
```

YaelDillies commented on Mathlib/GroupTheory/Exponent.lean:
```diff
@@ -213,7 +221,7 @@ theorem exponent_dvd_iff_forall_pow_eq_one {n : ℕ} : exponent G ∣ n ↔ ∀
+@[to_additive]
 theorem exponent_dvd_of_forall_pow_eq_one (n : ℕ) (hG : ∀ g : G, g ^ n = 1) :
     exponent G ∣ n :=
   (exponent_dvd_iff_forall_pow_eq_one G).mpr hG
```
Can you make that an `alias` of `exponent_dvd_iff_forall_pow_eq_one` (and make `G` implicit in `exponent_dvd_iff_forall_pow_eq_one`)?

Ruben-VandeVelde commented:
I've run leaff on this PR, though I don't fully trust the output:
```diff
+ added Monoid.exponent_pos.{u}
+ added Monoid.ExponentExists.isOfFinOrder.{u}
+ added AddMonoid.lcm_addOrderOf_eq_exponent.{u}
+ added AddMonoid.ExponentExists.exponent_pos.{u}
+ added AddMonoid.addOrderOf_le_exponent.{u}
+ added Monoid.exponent_dvd_iff_forall_pow_eq_one.{u}
+ added AddMonoid.exists_addOrderOf_eq_exponent.{u}
+ added AddMonoid.ExponentExists.isOfFinAddOrder.{u}
+ added Monoid.ExponentExists.orderOf_pos.{u}
+ added Monoid.ExponentExists.of_finite.{u}
+ added AddMonoid.exponent_dvd_of_forall_addOrderOf_dvd.{u}
+ added commMonoidOfExponentTwo.{u}
+ added Monoid.ExponentExists.exponent_ne_zero.{u}
+ added AddMonoid.ExponentExists.of_finite.{u}
+ added Group.exponent_dvd_nat_card.{u}
+ added AddMonoid.exponent_pos.{u}
+ added Monoid.exponent_dvd.{u}
+ added Monoid.orderOf_le_exponent.{u}
+ added Monoid.exponent_dvd_of_forall_orderOf_dvd.{u}
+ added Commute.of_orderOf_dvd_two.{u}
+ added AddMonoid.ExponentExists.exponent_ne_zero.{u}
+ added Group.exponent_dvd_card.{u}
+ added AddMonoid.exponent_ne_zero.{u}
+ added AddCommute.of_addOrderOf_dvd_two.{u}
+ added AddMonoid.ExponentExists.addOrderOf_pos.{u}
+ added addCommMonoidOfExponentTwo.{u}
+ added AddMonoid.exponent_dvd_iff_forall_nsmul_eq_zero.{u}
+ added Monoid.ExponentExists.exponent_pos.{u}
+ added AddMonoid.exponent_dvd.{u}
+ added Monoid.lcm_orderOf_eq_exponent.{u}
+ added Monoid.exists_orderOf_eq_exponent.{u}
+ added Monoid.exponent_ne_zero.{u}
! type changed for mul_not_mem_of_exponent_two
! type changed for mul_comm_of_exponent_two
! type changed for add_not_mem_of_addOrderOf_eq_two
! type changed for mul_not_mem_of_orderOf_eq_two
! type changed for inv_eq_self_of_exponent_two
! type changed for instCommGroupOfExponentTwo
! type changed for instAddCommGroupOfExponentTwo
! type changed for inv_eq_self_of_orderOf_eq_two
! type changed for add_comm_of_exponent_two
! type changed for neg_eq_self_of_addOrderOf_eq_two
! type changed for neg_eq_self_of_exponent_two
! type changed for orderOf_eq_two_iff
! type changed for add_not_mem_of_exponent_two
! type changed for addOrderOf_eq_two_iff
! type changed for AddMonoid.exponent_dvd_of_forall_nsmul_eq_zero
! type changed for Monoid.exponent_dvd_of_forall_pow_eq_one
+ attribute protected added to Monoid.ExponentExists.exponent_pos
+ attribute protected added to AddMonoid.ExponentExists.exponent_ne_zero
+ attribute reducible added to addCommMonoidOfExponentTwo
+ attribute reducible added to commMonoidOfExponentTwo
+ attribute protected added to AddMonoid.ExponentExists.exponent_pos
+ attribute protected added to Monoid.ExponentExists.exponent_ne_zero
! moved mul_not_mem_of_exponent_two from Mathlib.GroupTheory.SpecificGroups.KleinFour to Mathlib.GroupTheory.Exponent
! moved addOrderOf_eq_two_iff from Mathlib.GroupTheory.SpecificGroups.KleinFour to Mathlib.GroupTheory.Exponent
! moved add_not_mem_of_addOrderOf_eq_two from Mathlib.GroupTheory.SpecificGroups.KleinFour to Mathlib.GroupTheory.Exponent
! moved inv_eq_self_of_orderOf_eq_two from Mathlib.GroupTheory.SpecificGroups.KleinFour to Mathlib.GroupTheory.Exponent
! moved inv_eq_self_of_orderOf_eq_two from Mathlib.GroupTheory.SpecificGroups.KleinFour to Mathlib.GroupTheory.Exponent
! moved add_comm_of_exponent_two from Mathlib.GroupTheory.SpecificGroups.KleinFour to Mathlib.GroupTheory.Exponent
! moved add_comm_of_exponent_two from Mathlib.GroupTheory.SpecificGroups.KleinFour to Mathlib.GroupTheory.Exponent
! moved mul_comm_of_exponent_two from Mathlib.GroupTheory.SpecificGroups.KleinFour to Mathlib.GroupTheory.Exponent
! moved neg_eq_self_of_addOrderOf_eq_two from Mathlib.GroupTheory.SpecificGroups.KleinFour to Mathlib.GroupTheory.Exponent
! moved neg_eq_self_of_addOrderOf_eq_two from Mathlib.GroupTheory.SpecificGroups.KleinFour to Mathlib.GroupTheory.Exponent
! moved mul_not_mem_of_orderOf_eq_two from Mathlib.GroupTheory.SpecificGroups.KleinFour to Mathlib.GroupTheory.Exponent
! moved neg_eq_self_of_exponent_two from Mathlib.GroupTheory.SpecificGroups.KleinFour to Mathlib.GroupTheory.Exponent
! moved neg_eq_self_of_exponent_two from Mathlib.GroupTheory.SpecificGroups.KleinFour to Mathlib.GroupTheory.Exponent
! moved mul_not_mem_of_orderOf_eq_two from Mathlib.GroupTheory.SpecificGroups.KleinFour to Mathlib.GroupTheory.Exponent
! moved instCommGroupOfExponentTwo from Mathlib.GroupTheory.SpecificGroups.KleinFour to Mathlib.GroupTheory.Exponent
! moved instCommGroupOfExponentTwo from Mathlib.GroupTheory.SpecificGroups.KleinFour to Mathlib.GroupTheory.Exponent
! moved add_not_mem_of_exponent_two from Mathlib.GroupTheory.SpecificGroups.KleinFour to Mathlib.GroupTheory.Exponent
! moved add_not_mem_of_exponent_two from Mathlib.GroupTheory.SpecificGroups.KleinFour to Mathlib.GroupTheory.Exponent
! moved mul_comm_of_exponent_two from Mathlib.GroupTheory.SpecificGroups.KleinFour to Mathlib.GroupTheory.Exponent
! moved addOrderOf_eq_two_iff from Mathlib.GroupTheory.SpecificGroups.KleinFour to Mathlib.GroupTheory.Exponent
! moved inv_eq_self_of_exponent_two from Mathlib.GroupTheory.SpecificGroups.KleinFour to Mathlib.GroupTheory.Exponent
! moved orderOf_eq_two_iff from Mathlib.GroupTheory.SpecificGroups.KleinFour to Mathlib.GroupTheory.Exponent
! moved orderOf_eq_two_iff from Mathlib.GroupTheory.SpecificGroups.KleinFour to Mathlib.GroupTheory.Exponent
! moved inv_eq_self_of_exponent_two from Mathlib.GroupTheory.SpecificGroups.KleinFour to Mathlib.GroupTheory.Exponent
! moved add_not_mem_of_addOrderOf_eq_two from Mathlib.GroupTheory.SpecificGroups.KleinFour to Mathlib.GroupTheory.Exponent
! moved mul_not_mem_of_exponent_two from Mathlib.GroupTheory.SpecificGroups.KleinFour to Mathlib.GroupTheory.Exponent
! moved instAddCommGroupOfExponentTwo from Mathlib.GroupTheory.SpecificGroups.KleinFour to Mathlib.GroupTheory.Exponent
! moved instAddCommGroupOfExponentTwo from Mathlib.GroupTheory.SpecificGroups.KleinFour to Mathlib.GroupTheory.Exponent
```

YaelDillies reviewed (approved):
maintainer merge

j-loreaux commented:
bors merge
